shopfloor_reception/services/reception.py

```diff
@@ -302,91 +322,165 @@ def _scan_document__by_picking(self, barcode):
-        search = self._actions_for("search")
-        product = search.product_from_scan(barcode)
-        if product:
-            move = fields.first(
-                picking.move_lines.filtered(
-                    lambda m: m.product_id == product
-                    and float_compare(
-                        m.quantity_done,
-                        m.product_uom_qty,
-                        precision_rounding=m.product_uom.rounding,
-                    )
-                    == -1
-                )
+    def _scan_document__by_origin_move(self, moves, barcode):
+        if self.work.menu.allow_return:
+            pickings = moves.picking_id
+            outgoing_pickings = pickings.filtered(
+                lambda p: (p.picking_type_code == "outgoing")
             )
-            message = self._check_move_available(move, message_code="product")
-            if message:
-                return self._response_for_select_move(
-                    picking,
-                    message=message,
+            # If we find valid pickings for a return, then we create an empty
+            # return picking
+            if outgoing_pickings:
+                return_picking = self._scan_document__create_return(
+                    fields.first(outgoing_pickings), barcode
                 )
-            return self._select_line_from_product(picking, move, product)
+                return self._response_for_select_move(return_picking)
+        # A return picking has been scanned, but allow rma is disabled.
+        return self._scan_document__fallback()
+
+    def _scan_document__fallback(self):
+        return self._response_for_select_document(
+            message=self.msg_store.barcode_not_found()
+        )
 
-    def _scan_line__by_packaging(self, picking, barcode):
+    def _scan_line__create_return_move(self, return_picking, origin_moves):
+        # copied from odoo/src/addons/stock/wizard/stock_picking_return.py
+        stock = self._actions_for("stock")
+        return stock.create_return_move(return_picking, origin_moves)
+
+    def _scan_line__by_product__return(self, picking, product):
         search = self._actions_for("search")
-        packaging = search.packaging_from_scan(barcode)
-        if packaging:
-            move = fields.first(
-                picking.move_lines.filtered(
-                    lambda m: packaging in m.product_id.packaging_ids
-                    and float_compare(
-                        m.quantity_done,
-                        m.product_uom_qty,
-                        precision_rounding=m.product_uom.rounding,
-                    )
-                    == -1
-                )
+        origin_moves = search.origin_move_from_scan(picking.origin)
+        origin_moves = origin_moves.filtered(
+            lambda m: m.picking_id.picking_type_code == "outgoing"
+        )
+        origin_moves_for_product = origin_moves.filtered(
+            lambda m: m.product_id == product
+        )
```

Comment:
This is inefficient. Give a domain to origin_move_from_scan

Reply:
done